core/object/object.h

```diff
@@ -690,7 +699,19 @@ class Object {
 	}
 	virtual String get_save_class() const { return get_class(); } //class stored when saving
 
+	StringName get_script_class_name() const;
+
 	virtual bool is_class(const String &p_class) const {
+		if (!p_class.is_empty() && !script.is_null()) {
+			Object *s = script;
+			while (s) {
+				if (s->call(StringName("get_global_class_name")).operator String() == p_class) {
+					return true;
+				}
+				s = s->call(StringName("get_base_script"));
+			}
+		}
```

Comment:
While I believe that it would clearly make more *sense* for an `is_class` method to return the name of the best understood typename, as you have implemented, I also believe that modifying this method in this manner is too dangerous. `is_class()` is used frequently throughout the engine with the assumption that the returned class name will correspond to an actual natively-implemented class stored in the `ClassDB`. This makes it still work properly and be applicable for extensions, but not so for scripted types.

Rather than this sort of solution, I would *much* prefer to see something where we add an entirely different method to the base class that handles abstracting away the complexity of the script class name vs. the native class name. In which case, we should decide on the name of that method in the related Issue (#21789) probably. Other than that though, I like a lot of the modifications you've made, personally.

This also goes for the `is_class` method definition in the object macro further up in this same file.

Reply:
I'm confused as to why adding the Script type checking to `is_class` could be a bad thing, or why a separate method would be required? This method will still work if passed a Native or Extension type.

The only possible downside/regression I can see is if a developer were to shadow an Native or Extension type with a Script type, that false-positives could happen, but shadowing is bad practice anyways.

Thank you. :)

Reply:
> I'm confused as to why adding the Script type checking to is_class could be a bad thing

I think I initially confused this with `get_class()` for some reason which is used pretty extensively in the engine source code. I don't know how often `is_class()` is used, but if it's a lot less then *maybe* it would be less problematic? Still would end up adding to the inconsistency vs. the similarly named `get_class()` though.

It's less so that it's a "bad thing" if you added this functionality, but more so that you'd need to double check every reference to this method in the rest of the codebase to ensure it's logic is still consistent should the value return `true` unexpectedly. For example, if the code passed in a typename string to verify that the given typename is the name of an inherited *native* type when (cause it needs to verify it to do some native-only operation that is incompatible with scripted types), then making this change would suddenly make it possible for previously unsupported - and correctly incompatible - types to now slip through that check. That's the "risk". We would need to ensure that things that use this method to control flow logic don't e.g. do anything with the `ClassDB` since those operations would surely fail and/or log an error after this change is made.

> why a separate method would be required?

If a separate method is defined instead, then it means no one would have to do the above "cleanup" refactoring/investigating throughout the rest of the codebase after the change is made. BUT, I do agree that it could lead to redundancy / choice paralysis or confusion regarding which method should be used in which circumstances (a reoccurring source of confusion in Godot's API for newbies imo). So, it's not that it would be required. Just that IF we made the change, we'd need a corresponding correctional sweep of the rest of the codebase to go alongside it.

Reply:
>1st part
I can understand if this could cause some problems in places with Shadowing, but honestly I think this should be fixed on the GDScript side, in that `class_name` shouldn't be able to contain the name of an already existing Type. At least that seems like the more logical change personally anyways.
>2nd part
Instead of a separate method, I would instead offer a defaulted second parameter as to whether or not Script Types are considered.

Reply:
That sounds like a reasonable approach.

Reply:
> That sounds like a reasonable approach.

Okay, well I have a commit archived that I can push anytime before this gets merged if that is the route we wish to take.
However personally I would rather the Shadowing get fixed where it is being caused, so that the second parameter is not necessary , although that would need to be a separate PR to this one.

Reply:
So I just checked: https://github.com/godotengine/godot/blob/7686cfec64a9cbfe1c27bfa4f8b1eaf43755830e/modules/gdscript/gdscript_analyzer.cpp#L226

GDScript already checks for Shadowing and throws errors, which means I don't see a reason for this as is, to cause any issues.